### PR TITLE
Fix lifetime of scrolling hit object not updated when scrolling speed is adjusted

### LINKED ISSUE #12897: Scrolling HitObject lifetimes aren't updated correctly when scroll speed is adjusted
**Describe the bug:** 

Scrolling HitObjects have their lifetimes set by `ScrollingHitObjectContainer`. Initially they start out with `LifetimeStart  = StartTime - InitialLifetimeOffset(10s)`. When they become alive, a drawable will be assigned to them and added to the container, [then the newly added DHO will be invalidated to compute a better `LifetimeStart`](https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs#L159). This assignment of a new `LifetimeStart` will cause the HitObject to die, and DHO returned to pools. No further lifetime updates will be made until the HitObject becomes alive again(which may not be correct due to incorrect `LifetimeStart`), this causes problems with rulesets that allow the user to modify the scroll speed such as mania and Rush.

I wanted to move the lifetime calculation to a `ScrollingHitObjectLifetimeEntry` using a bindable, but found that [the LifetimeStart calculations used](https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs#L250-L276) depended on the draw properties of a `DrawableHitObject`, which the LifetimeEntries don't have access to. #12627 originally planned to provide a `GetDrawRectangle` which could be used to get a decent enough approximation, but that change has been reverted.

**Screenshots or videos showing encountered issue:** 

https://user-images.githubusercontent.com/12001167/119194539-85034c00-ba83-11eb-8b0e-652beb90906a.mp4

**osu!lazer version:** 
2021.515.0, 51e52fffd86347dbdb16c1420d1b3df13debfb1b on master (Pooled mania isn't in 2021.515.0, but the problem was already present then, affecting Rush)

**Logs:**
N/A

### PR DESCRIPTION
**Pull request by ekrctb**
- Fixes #12897
- [x] Depends on #13256
- [x] Depends on #13257 for merge conflict resolution (not strictly necessary)

The approach is different from my previous PRs.

- *Initial lifetime* is the `LifetimeStart` set in `HitObjectLifetimeEntry`. It should be a conservative estimation of the lifetime. If this lifetime was an underestimation (not covering the actual lifetime) then issues will occur (it was the case before this PR).
- More precise scrolling speed based `LifetimeStart` is set in `ScrollingHitObjectContainer`. This lifetime is computed when the hit object is alive, so the lifetime computation can use `DrawableHitObject` properties, such as the drawable size.

The critical observation is that it is okay to reset the lifetime to the initial lifetime at any time.

- If DHO was alive and the initial lifetime is set, it may live longer when the replay is rewound. No difference for normal gameplay.
- If DHO was not alive, the lifetime will be updated to the more precise one when the DHO becomes alive. If the DHO becomes immediately alive due to the longer lifetime, the lifetime will be updated in the next frame.

So I can just call `SetInitialLifetime` to all entries when the scrolling container is invalidated.

### CODE REVIEW

File: osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs [hunk 1]
```diff
@@ -94,6 +94,6 @@ internal bool KeepAlive
         /// <summary>
         /// Set <see cref="LifetimeEntry.LifetimeStart"/> using <see cref="InitialLifetimeOffset"/>.
         /// </summary>
-        private void setInitialLifetime() => LifetimeStart = HitObject.StartTime - InitialLifetimeOffset;
+        public void SetInitialLifetime() => LifetimeStart = HitObject.StartTime - InitialLifetimeOffset;
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs:97]: Could maybe be `internal`, unless there is a good reason for public exposition.

File: osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs [hunk 2]
```diff
@@ -150,81 +147,41 @@ private void flipPositionIfRequired(ref float position)
...
-
-        private void onAddRecursive(DrawableHitObject hitObject)
+        protected override void OnAdd(DrawableHitObject drawableHitObject)
         {
-            invalidateHitObject(hitObject);
-
-            hitObject.DefaultsApplied += invalidateHitObject;
-
-            foreach (var nested in hitObject.NestedHitObjects)
-                onAddRecursive(nested);
```
Comment (bdach) [inline osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:164]: Why was the recursion wrt nested objects removed here? I'm not seeing it explained anywhere, and it seems like a pretty big thing to remove.

Comment (ekrctb) [reply osu.Game/Rulesets/UI/Scrolling/ScrollingHitObjectContainer.cs:164]: This change is not part of this PR. It is included in #13257.

File: osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/HitObjectLifetimeEntry.cs]: I was about to say that I try to make sure anything we create can be replicated by a ruleset - so someone should be able to implement a SHOC themselves. But realised that `KeepAlive` is also internal, so... I've made the change.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**